HubSpec: MergeHub must immediately cancel new producers while draining

https://github.com/akka/akka/pull/30451/checks?check_run_id=3233793800

```
must immediately cancel new producers while draining *** FAILED *** (2 seconds, 41 milliseconds)
[info]   java.lang.AssertionError: assertion failed: expected class akka.stream.testkit.TestSubscriber$OnNext, found class akka.stream.testkit.TestSubscriber$OnComplete$ (OnComplete)
[info]   at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:572)
[info]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:543)
[info]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:159)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:973)
[info]   at akka.stream.testkit.TestSubscriber$ManualProbe.expectNextN(StreamTestKit.scala:438)
[info]   at akka.stream.scaladsl.HubSpec.f$proxy13$1$$anonfun$1(HubSpec.scala:218)
[info]   at akka.stream.testkit.scaladsl.StreamTestKit$.assertAllStagesStopped(StreamTestKit.scala:32)
[info]   at akka.stream.scaladsl.HubSpec.f$proxy13$2(HubSpec.scala:223)
[info]   at akka.stream.scaladsl.HubSpec.$init$$$anonfun$16$$anonfun$13(HubSpec.scala:202)```

In another GHACtions run (https://github.com/akka/akka/pull/30449/checks?check_run_id=3252611503#step:6:20256) a different test in HubSpec failed:

(yet another case of a different test on an already reported flaky suite)

```

[info] - must complete after draining control is invoked and all connected producers complete *** FAILED *** (2 seconds, 22 milliseconds)
[info]   java.lang.AssertionError: assertion failed: expected class akka.stream.testkit.TestSubscriber$OnNext, found class akka.stream.testkit.TestSubscriber$OnComplete$ (OnComplete)
[info]   at scala.Predef$.assert(Predef.scala:279)
[info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:572)
[info]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:543)
[info]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:542)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:973)
[info]   at akka.stream.testkit.TestSubscriber$ManualProbe.expectNextN(StreamTestKit.scala:438)
[info]   at akka.stream.scaladsl.HubSpec.$anonfun$new$29(HubSpec.scala:198)
[info]   at akka.stream.testkit.scaladsl.StreamTestKit$.assertAllStagesStopped(StreamTestKit.scala:32)
[info]   at akka.stream.scaladsl.HubSpec.$anonfun$new$28(HubSpec.scala:188)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1076)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
```

> In another GHACtions run (https://github.com/akka/akka/pull/30449/checks?check_run_id=3252611503#step:6:20256) a different test in HubSpec failed:
> 
> (yet another case of a different test on an already reported flaky suite)
> 
> ```
> 
> [info] - must complete after draining control is invoked and all connected producers complete *** FAILED *** (2 seconds, 22 milliseconds)
> [info]   java.lang.AssertionError: assertion failed: expected class akka.stream.testkit.TestSubscriber$OnNext, found class akka.stream.testkit.TestSubscriber$OnComplete$ (OnComplete)
> [info]   at scala.Predef$.assert(Predef.scala:279)
> [info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:572)
> [info]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:543)
> [info]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:542)
> [info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:973)
> [info]   at akka.stream.testkit.TestSubscriber$ManualProbe.expectNextN(StreamTestKit.scala:438)
> [info]   at akka.stream.scaladsl.HubSpec.$anonfun$new$29(HubSpec.scala:198)
> [info]   at akka.stream.testkit.scaladsl.StreamTestKit$.assertAllStagesStopped(StreamTestKit.scala:32)
> [info]   at akka.stream.scaladsl.HubSpec.$anonfun$new$28(HubSpec.scala:188)
> [info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
> [info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
> [info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
> [info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
> [info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
> [info]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1076)
> [info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
> ```

That indeed looks very similar, and likely the same fix has to be applied to that test.

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala
@@ -18,7 +18,7 @@ import akka.stream.testkit.Utils.TE
 import akka.stream.testkit.scaladsl.StreamTestKit._
 import akka.stream.testkit.scaladsl.TestSink
 import akka.stream.testkit.scaladsl.TestSource
-import akka.testkit.{ EventFilter, GHExcludeTest }
+import akka.testkit.EventFilter
 
 class HubSpec extends StreamSpec {
   implicit val ec: ExecutionContext = system.dispatcher
@@ -201,8 +201,7 @@ class HubSpec extends StreamSpec {
       downstream.expectComplete()
     }
 
-    // Excluded in GH Actions: https://github.com/akka/akka/issues/30464
-    "immediately cancel new producers while draining" taggedAs GHExcludeTest in assertAllStagesStopped {
+    "immediately cancel new producers while draining" in assertAllStagesStopped {
       val downstream = TestSubscriber.probe[Int]()
       val (sink, draining) =
         MergeHub.sourceWithDraining[Int](16).take(20).toMat(Sink.fromSubscriber(downstream))(Keep.left).run()

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala
@@ -4,10 +4,11 @@
 
 package akka.stream.scaladsl
 
-import scala.collection.immutable
-import scala.concurrent.Await
-import scala.concurrent.duration._
+import akka.Done
 
+import scala.collection.immutable
+import scala.concurrent.{ Await, ExecutionContext, Future, Promise }
+import scala.concurrent.duration._
 import akka.stream.KillSwitches
 import akka.stream.ThrottleMode
 import akka.stream.testkit.StreamSpec
@@ -20,6 +21,7 @@ import akka.stream.testkit.scaladsl.TestSource
 import akka.testkit.{ EventFilter, GHExcludeTest }
 
 class HubSpec extends StreamSpec {
+  implicit val ec: ExecutionContext = system.dispatcher
 
   "MergeHub" must {
 
@@ -206,21 +208,30 @@ class HubSpec extends StreamSpec {
         MergeHub.sourceWithDraining[Int](16).take(20).toMat(Sink.fromSubscriber(downstream))(Keep.left).run()
       Source(1 to 10).runWith(sink)
       Source(11 to 20).runWith(sink)
-      draining.drainAndComplete()
 
-      downstream.request(10)
-      val firstBatch = downstream.expectNextN(10)
+      // Give the sources time to fill up the producer buffers before completing (and cancelling the sources)
+      val testResult = Promise[Done]()
+      system.scheduler.scheduleOnce(200.millis) {
+        testResult.completeWith(Future {
+          draining.drainAndComplete()
 
-      val upstream = TestPublisher.probe[Int]()
-      Source.fromPublisher(upstream).runWith(sink)
-      upstream.expectCancellation()
+          downstream.request(10)
+          val firstBatch = downstream.expectNextN(10)
 
-      downstream.request(10)
-      val secondBatch = downstream.expectNextN(10)
+          val upstream = TestPublisher.probe[Int]()
+          Source.fromPublisher(upstream).runWith(sink)
+          upstream.expectCancellation()
 
-      (firstBatch ++ secondBatch).sorted should ===(1 to 20)
+          downstream.request(10)
+          val secondBatch = downstream.expectNextN(10)
 
-      downstream.expectComplete()
+          (firstBatch ++ secondBatch).sorted should ===(1 to 20)
+
+          downstream.expectComplete()
+          Done
+        })
+      }
+      testResult.future.futureValue
     }
   }
 

